### PR TITLE
jewel: core: enable rocksdb for filestore

### PR DESCRIPTION
**Pull request by jdurgin**
need to get ceph-kvstore-tool changes in master, and determine default rocksdb settings based on perf testing (e.g. multithreaded compaction is not enabled by default, but that's the whole point)

### CODE REVIEW

File: src/common/config_opts.h [hunk 1]
```diff
@@ -859,6 +859,18 @@ OPTION(rocksdb_db_paths, OPT_STR, "")   // path,size( path,size)*
...
+OPTION(rocksdb_bloom_bits_per_key, OPT_INT, 20) // Number of bits per key for RocksDB's bloom filters
+OPTION(rocksdb_cache_index_and_filter_blocks, OPT_BOOL, true) // Whether to cache indices and filters in block cache
+OPTION(rocksdb_cache_index_and_filter_blocks_with_high_priority, OPT_BOOL, true) // Whether to cache indicies and filters in the block cache with high priority
+OPTION(rocksdb_pin_l0_filter_and_index_blocks_in_cache, OPT_BOOL, true) // Whether to pin Level 0 indices and bloom filters in the block cache
+OPTION(rocksdb_index_type, OPT_STR, "binary_search") // Type of index for SST files: binary_search, hash_search, two_level
+OPTION(rocksdb_partition_filters, OPT_BOOL, false) // (experimental) partition SST index/filters into smaller blocks
+OPTION(rocksdb_metadata_block_size, OPT_INT, 4*1024) // The block size for index partitions (0 = rocksdb default)
+OPTION(rocksdb_perf, OPT_BOOL, false) // Enabling this will have 5-10% impact on performance for the stats collection
+OPTION(rocksdb_collect_compaction_stats, OPT_BOOL, false) //For rocksdb, this behavior will be an overhead of 5%~10%, collected only rocksdb_perf is enabled.
```
Comment (amitkumar50) [inline src/common/config_opts.h]: comments of `rocksdb_collect_compaction_stats` , `rocksdb_collect_extended_stats` and `rocksdb_collect_memory_stats` are same?

**Comment by liewegas** [comment]
:+1:

**Comment by smithfarm** [comment]
PendingReleaseNotes?

**Comment by jdurgin** [comment]
I'll add a release note, I also want to change the default rocksdb options for filestore pending some larger scale testing next week. At the very least we'll want more than 1 compaction thread.

**Comment by smithfarm** [comment]
@jdurgin With this, would it be possible to eliminate leveldb as a build dependency? Is this already possible in luminous?

**Comment by vumrao** [comment]
jewel backport #18850 is needed for this backport.

**Comment by smithfarm** [comment]
@jdurgin Needs rebase. Please remove DNM and add jewel-batch-1 when done, thanks!

**Comment by jdurgin** [comment]
retest this please

**Comment by smithfarm** [comment]
retest this please

**Comment by smithfarm** [comment]
@jdurgin I put together a new `wip-jewel-backports` integration branch containing a new crop of backports, but the build fails with

```
rocksdb/librocksdb.a(db_impl.o): In function `DumpMallocStats':
/build/ceph-10.2.10-437-gd060720/src/rocksdb/db/db_impl.cc:401: undefined reference to `malloc_stats_print'
collect2: error: ld returned 1 exit status
Makefile:16271: recipe for target 'ceph_objectstore_bench' failed
make[4]: *** [ceph_objectstore_bench] Error 1
make[4]: *** Waiting for unfinished jobs....
ar: `u' modifier ignored since `D' is the default (see `U')
rocksdb/librocksdb.a(db_impl.o): In function `DumpMallocStats':
/build/ceph-10.2.10-437-gd060720/src/rocksdb/db/db_impl.cc:401: undefined reference to `malloc_stats_print'
```

I pushed to Shaman a branch containing just latest jewel + this PR to verify whether this PR is implicated in the build failure.

**Comment by smithfarm** [comment]
@jdurgin False alarm. This PR builds fine.

**Comment by smithfarm** [comment]
@jdurgin Please rebase.

**Comment by smithfarm** [comment]
@jdurgin This PR was included in a rados run at http://tracker.ceph.com/issues/21742#note-25

Initially there were several failures similar to http://tracker.ceph.com/issues/22063 and http://tracker.ceph.com/issues/22064

After three re-runs these were narrowed down to what seems like a pretty reproducible assert in the radosbench task of a test in "rados/upgrade". I opened a new bug for that, due to the newly increased reproducibility: http://tracker.ceph.com/issues/22598

**Comment by jdurgin** [comment]
@liewegas still ok with this for jewel?

**Comment by liewegas** [comment]
yeah!

**Pull request merged**